src/DataObject/DataObjectDocument.php

```diff
@@ -636,4 +641,17 @@ public function onRemoveFromSearchIndexes(string $event): void
         }
     }
 
+    public function getChildDocuments(SiteTree $page): array
+    {
+        $docs = [];
+
+        foreach ($page->AllChildren() as $record) {
+            $document = DataObjectDocument::create($record);
+            $docs[$document->getIdentifier()] = $document;
+            $docs = array_merge($docs, $document->getDependentDocuments());
```

Comment:
Can calling `getDependentDocuments()` here ever result in an infinite loop?

Reply:
If my theory is correct, then the CMS will prevent us from this happening especially on a `SiteTree` which I used as a guide. 

https://github.com/silverstripe/silverstripe-cms/blob/668744e728154818873bc24cb0b8d45f73728b95/code/Model/SiteTree.php#L1752
The delete method on each child page executes on after delete as well which is kind of the same thing which is to recursively call the delete function.

My initial change was to account only the immediate children (`SiteTree`) of a certain page but @michalkleiner made a valid point to account other relationships on each child pages hence made the changes.

Reply:
Am thinking of asking other vendors to stress test this implementation since this issue came from them since they have more content.